addons/external-dns/gcp/credentials/main.tf

```diff
@@ -0,0 +1,25 @@
+data "terraform_remote_state" "infra" {
+  backend = "local"
+  config = {
+    path = "../../../../infra/gcp/terraform.tfstate.d/gcp-${var.tsb_mp["cluster_id"]}-${var.gcp_k8s_regions[tonumber(var.tsb_mp["cluster_id"])]}/terraform.tfstate"
+  }
+}
+
+locals {
+  dns_project = endswith(var.tsb_fqdn, ".gcp.cx.tetrate.info") ? "dns-terraform-sandbox" : data.terraform_remote_state.infra.outputs.project_id
+}
+
+resource "google_service_account" "external_dns" {
```

Comment:
to keep simple lets have this helper function as part of the module that is designated to create a dns subzone and related service account per given cluster and output the variables as part of the defined module, it will be identical across the clusters too.

Reply:
I considered the subzone approach, but it has several cons:

There is already a predefined shared zone that is already preconfigured and records for that zone need to land there. It is common to expect that users may want to use the same domain for the MP and for the services they deploy for a consistent experience.
Enabling the subzone creation would require us to grant IAM admin roles (at least in gcp) to any user using this repo, and probably this is not what we want. We may, but I didn't want to impose that big permissions grant on a shared project now, to get the functionality going.
Note that is users are using a private project in gcp this is already creating records in the local DNS zone of their project.

I think reusing the existing zone makes sense. In GCP we can already accommodate other zones: we can use the shared one, or create a new public one in the user's project or even a private one. We just have to align the FQN module of the other providers to how we do stuff in GCP, instead of hacking this one.

Reply:
let's not get centric to GCP (i.e. I will not list the reasons on a public repo ), the cons are you listing do not account very basic collisions (i.e. at least two clusters) across multiple deployments leveraging the common dns zone, on top GCP private zone is a corner case deployment as of now (and not extended across the multiple clouds, i.e. all the demos we do - do include multiple clouds not GCP is the least preferred option), also to keep in mind the demos we do - do use public zones so after can be easily reached and entertained by the customers as well. I agree on allowing to push external-dns to the common parent zone too, but it should not be a default option and we should implement more safe model to start, I can re-write the PR if you want. 

IAM is a necessary evil to fight which is very hard to avoid (and the permissions for DNS management for example under GCP project already includes it, for AWS and Azure)  as OpenShift deployment does require a dedicated sub-zone. 

We need to do a sub-zone to overcome:

1. naming collisions, i.e. "vmgateway" service under dns-zone1/demo-ignasi1 will be affected by a new deployment under the same zone by another deployment dns-zone1/demo-sergey1
2. keeping the underlay DNS infrastructure clean
3. follow external-dns common deployment patterns, i.e. sub-zone per the given cluster which also would be an easy to describe to the end consumer, and account multiple clusters

Reply:
1. I disagree with the collision's argument. If a user has N clusters is very understandable they won't be assigning the same DNS name to each VMgateway, so that shouldn't be an issue. Note that the VM gateway gets the DNS name from the config you set in the CP resource, and we can expect users to not put the same name there in all clusters (as that's a publicly resolvable name), so the argument of the name collision does not apply.
2. If the AWS and Azure providers are as flexible as GCP and support the notion of Private, Public and Shared zone, the DNS infra remains clean. Note that you're only talking about Private, and you're assimilating Public to Shared in your argument, but you're missing one. The Publc zone individually managed is a Public zone different than the Shared one. This already meets the use case you're requesting of keeping the DNS infra clean and isolated. It's just that AWS and Azure are not yet as flexible as Google; let's fix that instead of doing stuff that only applies to an add-on.
3. Given the above two points and that users already would have a way to choose between Private, Public (owned only by them) and Shared, this looks more like an improvement that could be perfectly implemented in another pull request but not a major blocker for this one.

Reply:
>and the permissions for DNS management for example under GCP project already includes it, for AWS and Azure

This would mean coupling the extenral-dns add-on to some internals of how the MP is deployed and the fact that it configures DNS. Coupling this is a very bad design. If at some point we change the internals of how we deploy the MP, we would break extenral-dns. breaking add-ons when some internals of the MP module change is not a good design, especially if they can be completely decoupled, which is how it is right now.